[Sema] Fix c23 not checking CheckBoolLikeConversion

pileghoff opened the pull request:
Fixes issue #79435 

Checks for implicit conversion into boolean was previously triggered by `CheckBoolLikeConversion` for C.
When `bool` as a keyword was introduced in C23, `CheckBoolLikeConversion` would no longer trigger when using `-std=c23`, but since logical operators and conditional statements still operate on scalar values, the checks for implicit conversion into bool were never triggered.

This fix changes `CheckBoolLikeConversion` to not return early for C23, even though it has support for bools.

Commit:
[Sema] Fix c23 not checking CheckBoolLikeConversion

shafik reviewed (commented):
Thank you for the Fix! Your summary should be a description of the problem and how your solution fixes that problem. You should also link to the github issue in the summary.

Questions about the changes in the PR should go as a comment to the PR.

Please add a release note and this will need tests as well.

pileghoff commented:
> Thank you for the Fix! Your summary should be a description of the problem and how your solution fixes that problem. You should also link to the github issue in the summary.
> 
> Questions about the changes in the PR should go as a comment to the PR.
> 
> Please add a release note and this will need tests as well.

Noted. I have updated the description.

With regards to testing, I think I get how to write them, but I'm still unsure about where I'm suppose to put them. Would a new file (`implicit-boolike-conversion.c`?) be appropriate? 

Is the PR requirements/process documented anywhere? I had a look at [here](https://llvm.org/docs/Contributing.html) and i was looking around if there was a template, but i have found nothing.

pileghoff commented:
I can also see i broke the `n3042.c` test. Stuff like `if (null_val) {}` now throws a warning, since `null_val` is a constant.

I would expect this warning to be okay in that scenario?

pileghoff commented:
For testing i have fixed the new warnings in `n3042.c` and i have expanded `warn-int-in-bool-context.c` to check c23 specefically (which was broken, but is now fixed) and i added a few extra cases to the test. I will take a look at the release notes later today.

AaronBallman commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -16134,7 +16134,7 @@ static void CheckConditionalOperator(Sema &S, AbstractConditionalOperator *E,
 /// Input argument E is a logical expression.
 static void CheckBoolLikeConversion(Sema &S, Expr *E, SourceLocation CC) {
-  if (S.getLangOpts().Bool)
+  if (S.getLangOpts().Bool && !S.getLangOpts().C23)
```
I think it would be good to add some comments here as to why we're checking for !C23 explicitly, WDYT?

pileghoff replied:
Sounds very reasonable

AaronBallman reviewed (commented):
The changes look good to me in general, though I'm waiting for precommit CI to finish before approving just to make sure we caught any other fallout from the changes in the test system.

pileghoff commented:
If i can get the build to pass (it seems to fail on unrelated tests?), everything should be ready for review.

AaronBallman reviewed (changes requested):
It looks like a bunch of formatting changes came in with SemaChecking.cpp. You should only be formatting the code that your patch needs to touch. We've got a script for that: https://clang.llvm.org/docs/ClangFormat.html#script-for-patch-reformatting or https://clang.llvm.org/docs/ClangFormat.html#git-integration

Can you back out the unrelated formatting changes?

AaronBallman reviewed (approved):
LGTM, thank you for the fix!

AaronBallman commented:
This passed precommit CI before and the only difference is that there are formatting changes removed, so I'm landing this before precommit CI finishes.
